pyomo/contrib/solver/solvers/highs.py

```diff
@@ -96,6 +96,121 @@ def update(self):
+        self.linear_coefs = linear_coefs
+        self.quadratic_coefs = quadratic_coefs
+        self.last_quadratic_coef_values = [value(i.expr) for i in self.quadratic_coefs]
+        # Store the quadratic coefficients in dictionary format
+        self.quad_coef_dict = {}
+        self._initialize_quad_coef_dict()
+        # Flag to force first update of quadratic coefficients
+        self._first_update = True
+
+    def _initialize_quad_coef_dict(self):
+        for coef in self.quadratic_coefs:
+            v1_ndx = coef.row_idx
+            v2_ndx = coef.col_idx
+            # Ensure we're storing the lower triangular part
+            row = max(v1_ndx, v2_ndx)
+            col = min(v1_ndx, v2_ndx)
+
+            coef_val = value(coef.expr)
+            # Adjust for diagonal elements
+            if v1_ndx == v2_ndx:
+                coef_val *= 2.0
+
+            self.quad_coef_dict[(row, col)] = coef_val
+
+    def update(self):
+        """
+        Update the quadratic objective expression.
+        """
+        needs_quadratic_update = self._first_update
+
+        self.constant.update()
+        for coef in self.linear_coefs:
+            coef.update()
+
+        for ndx, coef in enumerate(self.quadratic_coefs):
+            current_val = value(coef.expr)
+            if current_val != self.last_quadratic_coef_values[ndx]:
+                needs_quadratic_update = True
+
+                v1_ndx = coef.row_idx
+                v2_ndx = coef.col_idx
+                row = max(v1_ndx, v2_ndx)
+                col = min(v1_ndx, v2_ndx)
+
+                # Adjust the diagonal to match Highs' expected format
+                if v1_ndx == v2_ndx:
+                    current_val *= 2.0
+
+                self.quad_coef_dict[(row, col)] = current_val
+
+                self.last_quadratic_coef_values[ndx] = current_val
+
+        # If anything changed, rebuild and pass the Hessian
+        if needs_quadratic_update:
+            self._build_and_pass_hessian()
+            self._first_update = False
+
+    def _build_and_pass_hessian(self):
+        """Build and pass the Hessian to HiGHS in CSC format"""
+        if not self.quad_coef_dict:
+            return
+
+        dim = self.highs.getNumCol()
+
+        # Build CSC format for the lower triangular part
+        q_value = []
+        q_index = []
+        q_start = [0] * dim
+
+        sorted_entries = sorted(
+            self.quad_coef_dict.items(), key=lambda x: (x[0][1], x[0][0])
+        )
+
+        last_col = -1
+        for (row, col), val in sorted_entries:
+            while col > last_col:
+                last_col += 1
+                if last_col < dim:
+                    q_start[last_col] = len(q_value)
+
+            # Add the entry
+            q_index.append(row)
+            q_value.append(val)
+
+        while last_col < dim - 1:
+            last_col += 1
+            q_start[last_col] = len(q_value)
+
+        nnz = len(q_value)
```

Comment:
I'm not sure what `nnz` means here - can you please name the variable more explicitly?

Reply:
I have tried to name the variables in accordance with the highspy function definition:
    def passHessian(
        self,
        dim: int,
        nnz: int,
        format: int,
        q_start: numpy.ndarray[typing.Any, numpy.dtype[numpy.int32]],
        q_index: numpy.ndarray[typing.Any, numpy.dtype[numpy.int32]],
        q_value: numpy.ndarray[typing.Any, numpy.dtype[numpy.float64]],
    ) -> HighsStatus: ...

it's used a lot in scipy, their sparse matrices/arrays have a nnz attributes which returns the number of non-zero entries. Maybe num_nz is clearer (closer to the gurobi model attribute 'NumNZs') ? Happy to change it to whatever you think is clearer

Reply:
Nah, as I said, jsiirola told me about that standard. I don't like it but it's not _our_ standard.